### ISSUE TITLE
urxvt rendering issues with chosen emoji

### ISSUE THREAD

**Issue by anund** [maralorn/nix-output-monitor#80]
urxvt has ~poor support for emoji. Nom has picked some emoji I'm not sure it can support anytime soon. Considering the existing set may be be better outside the context of urxvt here is some documentation to help others with the issue. (issue is purely informational)

Reproducing:
Agreeing on a version of urxvt `nix-shell -p rxvt-unicode-unwrapped`
Agreeing on a font configuration `urxvt -fn "xft:DejaVu Sans Mono:size=14,xft:Unifont:Style=Regular:antialias=false"`
Printing the characters of interest `echo "⏵ ✔ ⏳ ⚠ ↓ ↑ ✔ ∅ ⏱ ∑"`

Debugging:
Printing the hex code of characters that don't render in the launched urxvt `printf "%x %x %x" \'⏵ \'⏳ \'⏱` (the hourglass just renders poorly with this font choice)
Getting a list of fonts that support these characters `fc-list ":charset=23f5 23f3 23f1"` (Unifont claims too)

Finding some alternatives `▸ ⌛ ⌚` (25b8 231b 231a). The arrow has the most possibility for alternatives. The hourglass is the a lower code point usually missing falling sand. It renders pooly in Unifont but does have nicer versions in other fonts. The clock is a lower code point that happens to be a watch. It prints poorly in Unifont but does have nicer versions in others that look like a clock rather than a watch.

```bash
  nixpkgs.overlays = [
    (self: super: {
      nix-output-monitor = super.nix-output-monitor.overrideAttrs( oldAttrs: {
        patches = [ ./nom.patch ];
      });
    })
  ];
```
```patch
diff --git a/lib/NOM/Print.hs b/lib/NOM/Print.hs
index c48e2a7..aa32700 100644
--- a/lib/NOM/Print.hs
+++ b/lib/NOM/Print.hs
@@ -78,16 +78,16 @@ down = "↓"
 up = "↑"
 
 -- | U+23F1 STOPWATCH
-clock = "⏱"
+clock = "⌚"
 
 -- | U+23F5 BLACK MEDIUM RIGHT-POINTING TRIANGLE
-running = "⏵"
+running = "▸"
 
 -- | U+2714 HEAVY CHECK MARK
 done = "✔"
 
 -- | U+23F8 DOUBLE VERTICAL BAR
-todo = "⏸"
+todo = "⌛"
 
 -- | U+26A0 WARNING SIGN
 warning = "⚠"
```
Font fishing package list.
```bash
  fonts.fonts = with pkgs; [ 
    arphic-ukai
    arphic-uming
    corefonts
    crimson
    dejavu_fonts 
    font-awesome
    freefont_ttf 
    inconsolata
    inter
    joypixels
    liberation_ttf
    manrope
    montserrat
    nerdfonts
    noto-fonts
    noto-fonts-cjk
    noto-fonts-emoji
    noto-fonts-extra
    openmoji-color
    powerline-fonts
    roboto
    roboto-mono
    source-han-sans-japanese
    source-han-sans-korean
    source-han-sans-simplified-chinese
    source-han-sans-traditional-chinese
    source-sans-pro
    symbola
    ttf_bitstream_vera
    twitter-color-emoji
    ubuntu_font_family
    unifont 
    wqy_microhei
    wqy_zenhei
  ];
  nixpkgs.config.joypixels.acceptLicense = true;
```

**Comment by davidak** [comment]
⌛ is bad, because it has a different meaning!

>A classic hourglass, all of its sand having flowed into the bottom bulb. Indicates time has passed.

⏳ stands for:

>A classic hourglass, its sand still flowing from the top to bottom bulb. Indicates time is still passing

which perfectly fit's the waiting here

**Comment by anund** [comment]
One renders an hourglass emoji correctly in the specific mentioned terminal in most fonts the other does not. It's a trade off and not an actual patch suggestion. If you're using the above combinations of things this lets things render nicely. :shrug:

**Comment by maralorn** [comment]
Yeah, it’s a bit sad, that the other hourglass seems to have better support. This is hard to get right. I will have a look into this, once I pick up development again (in about two months).

**Comment by tejing1** [comment]
For the record, there is a `rxvt-unicode-emoji` package variant, which includes patches that seem to fix this issue. The left side of `⏱` clips a bit, but otherwise it looks fine.
![2023-12-08-015017_275x50_scrot](https://github.com/maralorn/nix-output-monitor/assets/5663576/aa84ba43-88e7-406e-abe9-85b150281389)

**Comment by maralorn** [comment]
We changed the selection of emoji a while pack. Handling should be better now. Please reopen if this is still a problem.

**Comment by tejing1** [comment]
`clock`, `running`, and `todo` display as boxes for me in stock `rxvt-unicode`:

![2024-11-23-122514_1610x364_scrot](https://github.com/user-attachments/assets/7b3ff166-de75-48b9-9694-e1cda8da3dab)

Personally, I've worked around the issue by patching nom to use some nerdfont symbols, which `rxvt-unicode` displays fine (assuming you have nerdfonts):

```nix
pkgs.nix-output-monitor.overrideAttrs (old: { postPatch = old.postPatch or "" + ''
        sed -ie ${lib.escapeShellArg ''
          s/↓/\\xf072e/
          s/↑/\\xf0737/
          s/⏱/\\xf520/
          s/⏵/\\xf04b/
          s/✔/\\xf00c/
          s/⏸/\\xf04d/
          s/⚠/\\xf071/
          s/∅/\\xf1da/
          s/∑/\\xf04a0/
          ''} lib/NOM/Print.hs
      ''; });
```
The one for `average` probably isn't the best choice, but I can't be bothered...
